File: packages/python/goldenpipe/goldenpipe/pipeline.py
"""Pipeline -- thin wrapper over the engine layer."""
from __future__ import annotations

import re

import polars as pl

from goldenpipe.engine.registry import StageRegistry
from goldenpipe.engine.reporter import Reporter
from goldenpipe.engine.resolver import Resolver
from goldenpipe.engine.runner import Runner
from goldenpipe.models.config import PipelineConfig, StageSpec
from goldenpipe.models.context import PipeContext, PipeResult, PipeStatus

_TABLE_NAME_RE = re.compile(r"^[a-zA-Z_][a-zA-Z0-9_.]*$")


class Pipeline:
    """High-level pipeline orchestrator."""

    def __init__(
        self,
        config: PipelineConfig | None = None,
        registry: StageRegistry | None = None,
        identity_opts: dict | None = None,
    ) -> None:
        self._config = config
        self._registry = registry or StageRegistry()
        if registry is None:
            self._registry.discover()
        # v1.2: identity_opts is only meaningful in the auto-config path.
        # When the caller supplied an explicit PipelineConfig (YAML), the
        # YAML is authoritative and identity_opts is ignored.
        self._identity_opts = identity_opts
        # Set by _plan_config on each run; exposes the last plan-first
        # auto-config decision (rule_name, confidence, evidence) for
        # introspection. None until the brain has planned at least once.
        self._last_plan = None

    def run(
        self,
        source: str | None = None,
        df: pl.DataFrame | None = None,
        *,
        duckdb_con=None,
        duckdb_table: str | None = None,
    ) -> PipeResult:
        ctx = PipeContext()

        if duckdb_con is not None and duckdb_table is not None:
            # Engine-resident source (relocatable-stage contract, Phase C v2):
            # the data ORIGINATES in DuckDB, so ``ctx.frame`` starts as a lazy
            # ``DuckDBFrame`` and a remote stage right after pays NO ingress
            # crossing. It materializes to ``df`` only at the first LOCAL stage
            # (the Runner does that once), or never for an all-remote pipeline.
            from goldenpipe.models.frame import DuckDBFrame

            if not _TABLE_NAME_RE.match(duckdb_table):
                return PipeResult(
                    status=PipeStatus.FAILED,
                    source=f"duckdb:{duckdb_table}",
                    input_rows=0,
                    errors=[f"Invalid DuckDB table name: {duckdb_table!r}"],
                )
            try:
                rel = duckdb_con.sql(f"SELECT * FROM {duckdb_table}")  # noqa: S608 - name validated above
                ctx.frame = DuckDBFrame(rel)  # engine-resident, NOT materialized
                ctx.metadata["duckdb_con"] = duckdb_con
                ctx.metadata["source"] = f"duckdb:{duckdb_table}"
                # Row count via a scalar COUNT -- a cheap engine aggregate, NOT
                # the full DataFrame pull Phase C confines to egress.
                ctx.metadata["input_rows"] = duckdb_con.sql(
                    f"SELECT count(*) FROM {duckdb_table}"  # noqa: S608 - name validated above
                ).fetchone()[0]
            except Exception as e:
                return PipeResult(
                    status=PipeStatus.FAILED,
                    source=f"duckdb:{duckdb_table}",
                    input_rows=0,
                    errors=[f"Failed to load DuckDB table: {e}"],
                )
        elif df is not None:
            ctx.df = df
            ctx.metadata["source"] = "<DataFrame>"
            ctx.metadata["input_rows"] = len(df)
        elif source:
            try:
                ctx.df = pl.read_csv(source, ignore_errors=True, encoding="utf8-lossy")
                ctx.metadata["source"] = source
                ctx.metadata["input_rows"] = len(ctx.df)
            except Exception as e:
                return PipeResult(
                    status=PipeStatus.FAILED,
                    source=source or "",
                    input_rows=0,
                    errors=[f"Failed to load data: {e}"],
                )
        else:
            return PipeResult(
                status=PipeStatus.FAILED,
                source="",
                input_rows=0,
                errors=["No source file or DataFrame provided"],
            )

        config = self._config or self._plan_config(ctx)

        try:
            plan = Resolver.resolve(config, self._registry)
        except Exception as e:
            return PipeResult(
                status=PipeStatus.FAILED,
                source=ctx.metadata.get("source", ""),
                input_rows=ctx.metadata.get("input_rows", 0),
                errors=[f"Pipeline resolution failed: {e}"],
            )

        runner = Runner(registry=self._registry)
        stages = runner.run(plan, ctx)
        return Reporter.build(ctx, stages)

    def _plan_config(self, ctx: PipeContext) -> PipelineConfig:
        """Plan-first auto-config: profile the loaded context, run the rule
        table, and materialize the chosen plan into a PipelineConfig.

        This is the "brain" (parity with GoldenMatch's controller): the shape
        of the pipeline is DECIDED from the data + InferMap-inferred schema
        rather than being the fixed scan/transform/dedupe list. The portable
        decision core (``autoconfig_planner``) is kept free of Polars/Pydantic
        for the later ``goldenpipe-core`` Rust port; this method is the host
        glue bracket.
        """
        from goldenpipe.autoconfig_glue import plan_to_config, profile_context
        from goldenpipe.autoconfig_planner import plan_pipeline

        profile = profile_context(ctx)
        plan = plan_pipeline(profile)
        self._last_plan = plan
        return plan_to_config(
            plan,
            self._registry.list_all(),
            self._identity_opts,
        )

    def _auto_config(self) -> PipelineConfig:
        available = self._registry.list_all()
        stage_specs: list[StageSpec | str] = []
        for name in ["goldencheck.scan", "goldenflow.transform", "goldenmatch.dedupe"]:
            if name in available:
                stage_specs.append(StageSpec(use=name))
        # v1.2: when identity_opts is supplied and the stage is discoverable,
        # auto-append `goldenmatch.identity_resolve` after dedupe with the
        # opts as its stage_config. Otherwise stay backwards-compatible.
        if self._identity_opts and "goldenmatch.identity_resolve" in available:
            stage_specs.append(StageSpec(
                use="goldenmatch.identity_resolve",
                config={**self._identity_opts},
            ))
        return PipelineConfig(pipeline="auto", stages=stage_specs)

File: packages/typescript/goldenpipe/src/core/pipeline.ts
/**
 * Pipeline — high-level orchestrator + programmatic run helpers.
 * Port of goldenpipe/pipeline.py + goldenpipe/_api.py (the DataFrame paths).
 *
 * Operates on `Row[]`. The file-loading `run(source)` entry point lives in
 * `node/` (it needs `node:fs`); the edge-safe core only exposes the in-memory
 * `runDf` / `runStages` paths.
 *
 * Edge-safe: no `node:` imports.
 */

import type { PipeContext, PipelineConfig, PipeResult, Row, Stage } from "./models.js";
import { makePipeContext, makePipelineConfig, makeStageSpec, PipeStatus } from "./models.js";
import { StageRegistry } from "./engine/registry.js";
import { Resolver } from "./engine/resolver.js";
import { Runner } from "./engine/runner.js";
import { Reporter } from "./engine/reporter.js";
import { buildDefaultRegistry } from "./adapters/index.js";

const DEFAULT_STAGE_ORDER = [
  "goldencheck.scan",
  "goldenflow.transform",
  "goldenmatch.dedupe",
];

export interface PipelineOptions {
  config?: PipelineConfig | undefined;
  registry?: StageRegistry | undefined;
}

export class Pipeline {
  private readonly config: PipelineConfig | undefined;
  private readonly registry: StageRegistry;

  constructor(options?: PipelineOptions) {
    this.config = options?.config;
    // When the caller supplies a registry, use it as-is; otherwise build the
    // default suite registry (load + scan + transform + dedupe).
    this.registry = options?.registry ?? buildDefaultRegistry();
  }

  /** Run the pipeline on an array of rows. */
  async run(rows: readonly Row[], source = "<rows>"): Promise<PipeResult> {
    const ctx = makePipeContext({
      df: [...rows] as Row[],
      metadata: { source, input_rows: rows.length },
    });

    const config = this.config ?? this.autoConfig();

    let plan;
    try {
      plan = Resolver.resolve(config, this.registry);
    } catch (e) {
      const message = e instanceof Error ? e.message : String(e);
      return {
        status: PipeStatus.FAILED,
        source,
        inputRows: rows.length,
        stages: {},
        artifacts: {},
        skipped: [],
        errors: [`Pipeline resolution failed: ${message}`],
        reasoning: {},
        timing: {},
      };
    }

    const runner = new Runner(this.registry);
    const stages = await runner.run(plan, ctx);
    return Reporter.build(ctx, stages);
  }

  /** Build the default check→flow→dedupe config from the available stages. */
  private autoConfig(): PipelineConfig {
    const available = this.registry.listAll();
    const stages = DEFAULT_STAGE_ORDER.filter((name) => name in available).map((name) =>
      makeStageSpec(name),
    );
    return makePipelineConfig({ pipeline: "auto", stages });
  }
}

/**
 * Run a pipeline on an array of rows. Zero-config (default suite chain) or with
 * an explicit PipelineConfig. Port of `_api.run_df`.
 */
export async function runDf(
  rows: readonly Row[],
  config?: PipelineConfig,
  source = "<rows>",
): Promise<PipeResult> {
  const pipe = new Pipeline(config !== undefined ? { config } : {});
  return pipe.run(rows, source);
}

/**
 * Run specific stages programmatically against rows. Port of `_api.run_stages`.
 * The auto-prepended `load` stage is removed since rows are already supplied.
 */
export async function runStages(stages: readonly Stage[], rows: readonly Row[]): Promise<PipeResult> {
  const registry = new StageRegistry();
  for (const s of stages) registry.register(s);

  const config = makePipelineConfig({
    pipeline: "programmatic",
    stages: stages.map((s) => makeStageSpec(s.info.name)),
  });

  const ctx: PipeContext = makePipeContext({
    df: [...rows] as Row[],
    metadata: { source: "<programmatic>", input_rows: rows.length },
  });

  const plan = Resolver.resolve(config, registry);
  plan.stages = plan.stages.filter((s) => s.name !== "load");

  const runner = new Runner(registry);
  const results = await runner.run(plan, ctx);
  return Reporter.build(ctx, results);
}
